_posts/2020-02-17-eda-architecture.md

```diff
@@ -0,0 +1,181 @@
+
+Now, the messages *are* the API, the only business and technical reference.
+
+After much thinking and experiments, we came with these core design principles that are very important, and after 4 months of production use, we are very glad we complied with them from the start:
+
+### Business Services and Data Processing Services
+
+We make a distinction between Business Services and Data Processing Services (a.k.a. utility services) to cleanly separate business logic from data processing complexity.
+
+DP Services are expected to be  **pure, stateless, services** that provide some kind of algorithmic data processing (computations, transformations, …). Moreover, they are also context free: they should not depend on business knowledge, assumptions or external data sources. All they need to do their processing must be in the command itself. They should not have to query a tier to get more data. DP services are kind of *universal* libraries and can even be provided by tiers.
+
+Business Services implement the customers' workflows and only focus on business rules and requirements to orchestrate and implement the value addition upon our customers' audio and data. They make use of the DP services as a library for that. There are very specific to us: you would never want to externalize your business services.
+
+Business Services persist the data they produce and are their unique trusted source of truth.
+
+Business Services build and maintain their own customer configuration from events on the bus.
+
+All services must be idempotent, that is, if they receive twice the exactly same message, they must behave identically and produce the same outputs.
+
+### Commands, Results and Events
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages emitted by business services and telling the bus what happened.
+
+A business service owns the type of Events it emits. It knows nothing about the services who will process them. It subscribes to the types of Events it needs but knows nothing about their origins.
+
+An Event type defines the meaning of the events of that type and their data schema. They must be documented.
+
+The type of an actual event message is given by its name (used as *routing key* in AMQP, see *Topology* below). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
```

Comment:
AMQP comes out of nowhere here. Maybe we could write something like:
```suggestion
The type of an actual event message is given by its name (used as *routing key* in [RabbitMQ AMQP protocol](https://www.rabbitmq.com/protocol.html), see *Topology* below). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
```
Or we could even avoid mentioning it: "(used as *routing key*, see *Topology* below)"